pkg/services/ngalert/migration/service.go

```diff
@@ -0,0 +1,200 @@
+			seenUIDs: uidSet{set: make(map[string]struct{}), caseInsensitive: ms.store.GetDialect().SupportEngine()},
+			silences: make(map[int64][]*pb.MeshSilence),
+			log:      ms.log,
+			dialect:  ms.store.GetDialect(),
+			cfg:      ms.cfg,
+			store:    ms.store,
+			sess:     sess.Session,
+		}
+
+		err := mg.Exec()
+		if err != nil {
+			return fmt.Errorf("executing migration: %w", err)
+		}
+
+		err = ms.SetMigrated(ctx, true)
+		if err != nil {
+			return fmt.Errorf("setting migration status: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("migration failed: %w", err)
+	}
+
+	return nil
+}
+
+// IsDisabled returns true if the cfg is nil.
+func (ms *MigrationService) IsDisabled() bool {
+	return ms.cfg == nil
+}
+
+func (ms *MigrationService) GetMigrated(ctx context.Context) (bool, error) {
+	content, exists, err := ms.kv.Get(ctx, migratedKey)
+	if err != nil {
+		return false, err
+	}
+
+	if !exists {
+		return false, nil
+	}
+
+	return strconv.ParseBool(content)
+}
+
+func (ms *MigrationService) SetMigrated(ctx context.Context, migrated bool) error {
+	return ms.kv.Set(ctx, migratedKey, strconv.FormatBool(migrated))
+}
+
+func (ms *MigrationService) Revert(ctx context.Context) error {
+	return ms.store.WithTransactionalDbSession(ctx, func(dbSess *db.Session) error {
+		sess := dbSess.Session
+		_, err := sess.Exec("delete from alert_rule")
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from alert_rule_version")
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from dashboard_acl where dashboard_id IN (select id from dashboard where created_by = ?)", FOLDER_CREATED_BY)
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from dashboard where created_by = ?", FOLDER_CREATED_BY)
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from alert_configuration")
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from ngalert_configuration")
+		if err != nil {
+			return err
+		}
+
+		_, err = sess.Exec("delete from alert_instance")
+		if err != nil {
+			return err
+		}
+
+		exists, err := sess.IsTableExist("kv_store")
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			_, err = sess.Exec("delete from kv_store where namespace = ?", notifier.KVNamespace)
+			if err != nil {
+				return err
+			}
+		}
+
+		files, err := getSilenceFileNamesForAllOrgs(ms.cfg.DataPath)
+		if err != nil {
+			return err
+		}
+		for _, f := range files {
+			if err := os.Remove(f); err != nil {
+				ms.log.Error("alert migration error: failed to remove silence file", "file", f, "err", err)
+			}
+		}
+
+		// Since both revert and this modifies kv_store, we use the same transaction in both writes to prevent Lock Wait Timeout.
+		withTransaction := context.WithValue(ctx, sqlstore.ContextSessionKey{}, dbSess)
+		err = ms.SetMigrated(withTransaction, false)
```

Comment:
I think propagation of the context would be a better choice. The framework will reuse the transaction from the context

Reply:
Agreed, but `WithTransactionalDbSession` doesn't store the session in context afaik, hence why I added it manually. Possibly a bug?

Reply:
Ah nvm, should call `InTransaction` first and then next the `WithTransactionalDbSession` :+1: